### PR TITLE
compiler wrapper: prioritize spack store paths in -L, -I, -rpath

### LINKED ISSUE #25832: Installation issue: gperftools
### Steps to reproduce the issue

```console
spack install gperftools
```

### Information on your system

* **Spack:** 0.16.2-3914-fbed679dd0
* **Python:** 3.9.6
* **Platform:** linux-ubuntu20.04-skylake
* **Concretizer:** original

### Additional information

Basically, this hits https://github.com/gperftools/gperftools/issues/424 because gperftools isn't configured to link with libunwind. Forcing a LDFLAGS="-L/path/to/libunwind -lunwind' bypasses the issue but it's an ugly solution. gperftools has cmake build now and we should switch the package to cmake build
@albestro  @eschnett  @msimberg  @teonnik

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

### PR DESCRIPTION
**Pull request by haampie**
Closes #25832

This commit partitions search paths of `-L`, `-I` (and `-rpath`) into three
groups, from highest priority to lowest:

1. Spack managed directories: these include (a) absolute paths prefixed
   by the path to the store, any upstream or the stage dir, and (b)
   relative paths, since they should be relative to the stage dir.
2. Non-system dirs: these are for externals that live in non-system
   locations, and are less likely to shadow other dependencies
3. System dirs: your typical `/usr/lib` etc, which is very likely to
   shadow other deps

It's very easy for Spack to know the prefixes it owns, it's much more
difficult to tell system dirs from non-system dirs. Before this commit
Spack tried to distinguish only system and non-system dirs, and failed
for very trivial cases like `/usr/lib/x/..` which comes up often, since
build systems sometimes copy search paths from `gcc -print-search-dirs`.
One example is the autotools build system of `gperftools`, where we
could not prevent the build system from linking to system libunwind
instead of the Spack dependency.

Potentially this implementation of path partitioning is faster than the
current state of things, since a loop over paths is replaced with an
eval'ed `case ...`:

```sh
path_order() {
case "$1" in
    $SPACK_MANAGED_DIRS) return 0 ;;
    $SPACK_SYSTEM_DIRS) return 2 ;;
    /*) return 1 ;;
esac
return 0
}
```

After this PR I can successfully install `spack install gperftools`
`build_system=autotools +libunwind` (after removing the conflict).

### CODE REVIEW

**Comment by haampie** [comment]
@trws mentioned that to distinguish group 2 and 3 we could look at the compiler's implicit link dir list instead of hard-coded `/usr/lib` etc. That sounds reproducible enough to me. Can be done as a follow-up PR.

**Comment by haampie** [comment]
This also slightly speeds up the compiler wrapper as a loop is replaced with a `case`. The following:

```
Benchmark 1: /home/harmen/spack/lib/spack/env/cc -L/home/harmen/spack/opt/spack/linux-ubuntu23.10-zen2/clang-16.0.6/libxml2-2.10.3-vlw4xiwlwxjppq36b35gqprkp4ohs3rg/lib -I/home/harmen/spack/opt/spack/linux-ubuntu23.10-zen2/clang-16.0.6/gettext-0.22.4-c2x4o6irrj7cawiteliu5jielyux6ytq/include -I/home/harmen/spack/opt/spack/linux-ubuntu23.10-zen2/clang-16.0.6/ncurses-6.4-45hgowckndfka3mk5hbl5kifnoj4qwbz/include -I/home/harmen/spack/opt/spack/linux-ubuntu23.10-zen2/clang-16.0.6/libxml2-2.10.3-vlw4xiwlwxjppq36b35gqprkp4ohs3rg/include -I/home/harmen/spack/opt/spack/linux-ubuntu23.10-zen2/clang-16.0.6/sqlite-3.43.2-syfc42zazods3tuj7sudy2f6d6rcogeb/include -L/home/harmen/spack/opt/spack/linux-ubuntu23.10-zen2/clang-16.0.6/util-linux-uuid-2.38.1-rtrv4i4fp7y6o7rfyn44mz2xd6rs3vdx/lib -Wl,-rpath,/home/harmen/spack/opt/spack/linux-ubuntu23.10-zen2/clang-16.0.6/gettext-0.22.4-c2x4o6irrj7cawiteliu5jielyux6ytq/lib -I/home/harmen/spack/opt/spack/linux-ubuntu23.10-zen2/clang-16.0.6/libbsd-0.12.1-kg277b55erksfdqijrj7b44lt6smgiav/include -I/home/harmen/spack/opt/spack/linux-ubuntu23.10-zen2/clang-16.0.6/readline-8.2-7o75xh5vmjosrvqpzsyphrt2ptq2ak2e/include -Wl,-rpath,/home/harmen/spack/opt/spack/linux-ubuntu23.10-zen2/clang-16.0.6/readline-8.2-7o75xh5vmjosrvqpzsyphrt2ptq2ak2e/lib -L/home/harmen/spack/opt/spack/linux-ubuntu23.10-zen2/clang-16.0.6/libiconv-1.17-cqfgo7darogzhwcjiymk5nq3d445oxnj/lib -Wl,-rpath,/home/harmen/spack/opt/spack/linux-ubuntu23.10-zen2/clang-16.0.6/bzip2-1.0.8-pomspdxbb6poodos52ylzpokxscl6cq2/lib -L/home/harmen/spack/opt/spack/linux-ubuntu23.10-zen2/clang-16.0.6/libffi-3.4.6-sifmtspocebiwpdoxploxvqbw6pbiycy/lib -I/home/harmen/spack/opt/spack/linux-ubuntu23.10-zen2/clang-16.0.6/util-linux-uuid-2.38.1-rtrv4i4fp7y6o7rfyn44mz2xd6rs3vdx/include -L/home/harmen/spack/opt/spack/linux-ubuntu23.10-zen2/clang-16.0.6/libxcrypt-4.4.35-wdmqjbecj4fxunmlhtx62b6qbt2lrpuv/lib -L/home/harmen/spack/opt/spack/linux-ubuntu23.10-zen2/clang-16.0.6/gdbm-1.23-zxcn33zqcrnmvm6k2maj374vmybtqyvt/lib -I/home/harmen/spack/opt/spack/linux-ubuntu23.10-zen2/clang-16.0.6/libffi-3.4.6-sifmtspocebiwpdoxploxvqbw6pbiycy/include -I/home/harmen/spack/opt/spack/linux-ubuntu23.10-zen2/clang-16.0.6/gdbm-1.23-zxcn33zqcrnmvm6k2maj374vmybtqyvt/include -Wl,-rpath,/home/harmen/spack/opt/spack/linux-ubuntu23.10-zen2/clang-16.0.6/libffi-3.4.6-sifmtspocebiwpdoxploxvqbw6pbiycy/lib -Wl,-rpath,/home/harmen/spack/opt/spack/linux-ubuntu23.10-zen2/clang-16.0.6/libxml2-2.10.3-vlw4xiwlwxjppq36b35gqprkp4ohs3rg/lib -L/home/harmen/spack/opt/spack/linux-ubuntu23.10-zen2/clang-16.0.6/expat-2.6.2-mgp2n357lbg444zzldn6fvuhi46vfotw/lib -I/home/harmen/spack/opt/spack/linux-ubuntu23.10-zen2/clang-16.0.6/libiconv-1.17-cqfgo7darogzhwcjiymk5nq3d445oxnj/include -L/home/harmen/spack/opt/spack/linux-ubuntu23.10-zen2/clang-16.0.6/gettext-0.22.4-c2x4o6irrj7cawiteliu5jielyux6ytq/lib -Wl,-rpath,/home/harmen/spack/opt/spack/linux-ubuntu23.10-zen2/clang-16.0.6/libbsd-0.12.1-kg277b55erksfdqijrj7b44lt6smgiav/lib -L/home/harmen/spack/opt/spack/linux-ubuntu23.10-zen2/clang-16.0.6/openssl-3.2.1-vnns5mnn2tcjml6umczthhijf6mhqdcj/lib64 -L/home/harmen/spack/opt/spack/linux-ubuntu23.10-zen2/clang-16.0.6/ncurses-6.4-45hgowckndfka3mk5hbl5kifnoj4qwbz/lib -L/home/harmen/spack/opt/spack/linux-ubuntu23.10-zen2/clang-16.0.6/libbsd-0.12.1-kg277b55erksfdqijrj7b44lt6smgiav/lib -Wl,-rpath,/home/harmen/spack/opt/spack/linux-ubuntu23.10-zen2/clang-16.0.6/python-3.12.1-eu5igwmh6zfqq3gh3g5laktolblkov5x/lib64 -Wl,-rpath,/home/harmen/spack/opt/spack/linux-ubuntu23.10-zen2/clang-16.0.6/sqlite-3.43.2-syfc42zazods3tuj7sudy2f6d6rcogeb/lib -Wl,-rpath,/home/harmen/spack/opt/spack/linux-ubuntu23.10-zen2/clang-16.0.6/ncurses-6.4-45hgowckndfka3mk5hbl5kifnoj4qwbz/lib -I/home/harmen/spack/opt/spack/linux-ubuntu23.10-zen2/clang-16.0.6/libxml2-2.10.3-vlw4xiwlwxjppq36b35gqprkp4ohs3rg/include/libxml2 -I/home/harmen/spack/opt/spack/linux-ubuntu23.10-zen2/clang-16.0.6/expat-2.6.2-mgp2n357lbg444zzldn6fvuhi46vfotw/include -L/home/harmen/spack/opt/spack/linux-ubuntu23.10-zen2/clang-16.0.6/xz-5.4.6-xxshcl5yvxb4y2zt25nco3ur2yq3r7nq/lib -I/home/harmen/spack/opt/spack/linux-ubuntu23.10-zen2/clang-16.0.6/zlib-ng-2.1.6-xidpmzycb2jufhr6doa7anltbcyhu263/include -L/home/harmen/spack/opt/spack/linux-ubuntu23.10-zen2/clang-16.0.6/zlib-ng-2.1.6-xidpmzycb2jufhr6doa7anltbcyhu263/lib -L/home/harmen/spack/opt/spack/linux-ubuntu23.10-zen2/clang-16.0.6/readline-8.2-7o75xh5vmjosrvqpzsyphrt2ptq2ak2e/lib -Wl,-rpath,/home/harmen/spack/opt/spack/linux-ubuntu23.10-zen2/clang-16.0.6/libxcrypt-4.4.35-wdmqjbecj4fxunmlhtx62b6qbt2lrpuv/lib -I/home/harmen/spack/opt/spack/linux-ubuntu23.10-zen2/clang-16.0.6/openssl-3.2.1-vnns5mnn2tcjml6umczthhijf6mhqdcj/include -I/home/harmen/spack/opt/spack/linux-ubuntu23.10-zen2/clang-16.0.6/xz-5.4.6-xxshcl5yvxb4y2zt25nco3ur2yq3r7nq/include -Wl,-rpath,/home/harmen/spack/opt/spack/linux-ubuntu23.10-zen2/clang-16.0.6/python-3.12.1-eu5igwmh6zfqq3gh3g5laktolblkov5x/lib -I/home/harmen/spack/opt/spack/linux-ubuntu23.10-zen2/clang-16.0.6/libxcrypt-4.4.35-wdmqjbecj4fxunmlhtx62b6qbt2lrpuv/include -L/home/harmen/spack/opt/spack/linux-ubuntu23.10-zen2/clang-16.0.6/libmd-1.0.4-ea4pc7pkxy7gcoobm5urzak7xm5hstt5/lib -Wl,-rpath,/home/harmen/spack/opt/spack/linux-ubuntu23.10-zen2/clang-16.0.6/expat-2.6.2-mgp2n357lbg444zzldn6fvuhi46vfotw/lib -I/home/harmen/spack/opt/spack/linux-ubuntu23.10-zen2/clang-16.0.6/libmd-1.0.4-ea4pc7pkxy7gcoobm5urzak7xm5hstt5/include -L/home/harmen/spack/opt/spack/linux-ubuntu23.10-zen2/clang-16.0.6/sqlite-3.43.2-syfc42zazods3tuj7sudy2f6d6rcogeb/lib -L/home/harmen/spack/opt/spack/linux-ubuntu23.10-zen2/clang-16.0.6/bzip2-1.0.8-pomspdxbb6poodos52ylzpokxscl6cq2/lib -Wl,-rpath,/home/harmen/spack/opt/spack/linux-ubuntu23.10-zen2/clang-16.0.6/zlib-ng-2.1.6-xidpmzycb2jufhr6doa7anltbcyhu263/lib -Wl,-rpath,/home/harmen/spack/opt/spack/linux-ubuntu23.10-zen2/clang-16.0.6/gdbm-1.23-zxcn33zqcrnmvm6k2maj374vmybtqyvt/lib -Wl,-rpath,/home/harmen/spack/opt/spack/linux-ubuntu23.10-zen2/clang-16.0.6/openssl-3.2.1-vnns5mnn2tcjml6umczthhijf6mhqdcj/lib64 -Wl,-rpath,/home/harmen/spack/opt/spack/linux-ubuntu23.10-zen2/clang-16.0.6/util-linux-uuid-2.38.1-rtrv4i4fp7y6o7rfyn44mz2xd6rs3vdx/lib -Wl,-rpath,/home/harmen/spack/opt/spack/linux-ubuntu23.10-zen2/clang-16.0.6/libiconv-1.17-cqfgo7darogzhwcjiymk5nq3d445oxnj/lib -Wl,-rpath,/home/harmen/spack/opt/spack/linux-ubuntu23.10-zen2/clang-16.0.6/xz-5.4.6-xxshcl5yvxb4y2zt25nco3ur2yq3r7nq/lib -Wl,-rpath,/home/harmen/spack/opt/spack/linux-ubuntu23.10-zen2/clang-16.0.6/libmd-1.0.4-ea4pc7pkxy7gcoobm5urzak7xm5hstt5/lib -I/home/harmen/spack/opt/spack/linux-ubuntu23.10-zen2/clang-16.0.6/bzip2-1.0.8-pomspdxbb6poodos52ylzpokxscl6cq2/include -O3 -fPIC -fstack-protector-strong -fno-strict-aliasing -fasynchronous-unwind-tables -Wall -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -Wno-unused-result -g -ffast-math -I.. -I../.. -I../lib -L. -L../lib -L../lib/.libs
```

runs in 17.0ms instead of 19.1ms

File: lib/spack/spack/test/build_environment.py [hunk 1]
```diff
@@ -63,7 +63,8 @@ def build_environment(working_env):
     os.environ["SPACK_LINKER_ARG"] = "-Wl,"
     os.environ["SPACK_DTAGS_TO_ADD"] = "--disable-new-dtags"
     os.environ["SPACK_DTAGS_TO_STRIP"] = "--enable-new-dtags"
-    os.environ["SPACK_SYSTEM_DIRS"] = "/usr/include /usr/lib"
+    os.environ["SPACK_SYSTEM_DIRS"] = "/usr/include|/usr/lib"
+    os.environ["SPACK_MANAGED_DIRS"] = f"{prefix}/opt/spack"
```
Comment (trws) [inline lib/spack/spack/test/build_environment.py:67]: This should probably be set to the value of `install_tree:root` from the config.

**Review by trws — CHANGES_REQUESTED** [review]
This looks like a good improvement, and the implementation is really clean (always tricky in the wrapper script). Just the one issue with managed dirs and this should be good to go.

File: lib/spack/spack/test/build_environment.py (hunk 1 above)
Comment (haampie) [reply lib/spack/spack/test/build_environment.py:67]: this is in a test fixture only used locally in unit tests here, no integration tests, so should be OK like this

Comment (trws) [reply lib/spack/spack/test/build_environment.py:67]: Got it, missed the context.  Thanks!

**Review by trws — APPROVED** [review]
LGTM!

**Comment by alecbcs** [comment]
@trws @haampie I can't seem to get rust to build on MacOS or Linux after this PR/commit. If I force using a `libssh2` that was built previous to this change I don't have any issue, but building up fresh after this change fails with the following error.

```
  >> 8489      cargo:warning=libssh2/src/kex.c:1473:50: error: use of undeclared identifier 'LIBSSH2_DH_GEX_OPTGROUP'
     8490
     8491      cargo:warning=        _libssh2_htonu32(key_state->request + 5, LIBSSH2_DH_GEX_OPTGROUP);
     8492
     8493      cargo:warning=                                                 ^
     8494
  >> 8495      cargo:warning=libssh2/src/kex.c:1474:50: error: use of undeclared identifier 'LIBSSH2_DH_GEX_MAXGROUP'
     8496
     8497      cargo:warning=        _libssh2_htonu32(key_state->request + 9, LIBSSH2_DH_GEX_MAXGROUP);
     8498
     8499      cargo:warning=                                                 ^
     8500
     8501      cargo:warning=libssh2/src/kex.c:2322:9: warning: 'EC_KEY_free' is deprecated [-Wdeprecated-declarations]
```

**Comment by trws** [comment]
I'm guessing that has to be a slicing error of some kind, does part of the error/warning have the paths of the headers that got selected so we can see what it picked?

**Comment by haampie** [comment]
I guess rust vendors libssh2 and compiles it outside of the stage dir using absolute paths in `-I <vendored libssh>`, and the compiler wrapper prioritizes `-I <spack dir>` over it, which happens to be shadowing the vendored copy of the headers.

#43797 ensures that rust doesn't build libssh2 when building cargo (I guess)

**Comment by adamjstewart** [comment]
I suspect this also broke torchtext in CI. Although @haampie tried to trigger a full pipeline rebuild, it looks like this didn't actually work: https://gitlab.spack.io/spack/spack/-/pipelines/677691

**Comment by haampie** [comment]
General fix in #43844

**Pull request merged**